### PR TITLE
Fix BinOp `ty()` assertion and `fn_sig()` for closures

### PR DESCRIPTION
**Pull request by celinval**
`BinOp::ty()` was asserting that the argument types were primitives. However, the primitive check doesn't include pointers, which can be used in a `BinaryOperation`. Thus extend the arguments to include them.

Since I had to add methods to check for pointers in TyKind, I just went ahead and added a bunch more utility checks that can be handy for our users and fixed the `fn_sig()` method to also include closures.

@compiler-errors just wanted to confirm that today no `BinaryOperation` accept SIMD types. Is that correct?

r? @compiler-errors

### CODE REVIEW

**Review by compiler-errors — CHANGES_REQUESTED** [review]
Some questions.

> just wanted to confirm that today no BinaryOperation accept SIMD types. Is that correct?

I don't think we lower SIMD intrinsics to MIR binops like we do for primitives, I can see it's not happening in `compiler/rustc_mir_transform/src/lower_intrinsics.rs`.

File: compiler/stable_mir/src/ty.rs [hunk 1]
```diff
@@ -259,6 +283,102 @@ impl TyKind {
         )
     }
 
+    /// A scalar type is one that denotes an atomic datum, with no sub-components.
+    /// (A RawPtr is scalar because it represents a non-managed pointer, so its
+    /// contents are abstract to rustc.)
+    #[inline]
+    pub fn is_scalar(&self) -> bool {
```
[the remark this answers is not in the document]
Comment (celinval) [reply compiler/stable_mir/src/ty.rs:290]: This is meant to be helper functions for tools developed on the top of StableMIR. In Kani, we use a bunch of these functions today, and while migrating to StableMIR, we have been writing a bunch of them. Like here: https://github.com/model-checking/kani/blob/9090db33a7d98e3ec25ec604d4f4cf9481fd6add/kani-compiler/src/codegen_cprover_gotoc/codegen/place.rs#L784-L790


 That said, I can definitely remove the ones that are not very clear like this one.

[the remark this answers is not in the document]
Comment (compiler-errors) [reply compiler/stable_mir/src/ty.rs:290]: I would remove this one in particular unless there is a particular use for it, or else it needs to have its behavior documented pretty precisely.

File: compiler/stable_mir/src/mir/body.rs [hunk 2]
```diff
@@ -278,8 +278,10 @@ impl BinOp {
     /// Return the type of this operation for the given input Ty.
     /// This function does not perform type checking, and it currently doesn't handle SIMD.
     pub fn ty(&self, lhs_ty: Ty, rhs_ty: Ty) -> Ty {
-        assert!(lhs_ty.kind().is_primitive());
-        assert!(rhs_ty.kind().is_primitive());
+        let lhs_kind = lhs_ty.kind();
+        let rhs_kind = rhs_ty.kind();
+        assert!(lhs_kind.is_primitive() || lhs_kind.is_any_ptr());
```
[the remark this answers is not in the document]
Comment (celinval) [reply compiler/stable_mir/src/mir/body.rs:283]:
```rust
/// Example of a function pointer being using in `BinOp::Eq`.
pub fn test_reify_fn_pointer() {
    assert!(poly::<usize> as fn() == poly::<usize> as fn());
}

fn poly<T>() {}

/// Here is an example of raw pointer being used with `BinOp::Le`:
pub fn check_raw_ptr() {
    let arr = [0, 1];
    let ptr_0 : *const u8 = &arr[0];
    let ptr_1 : *const u8 = &arr[1];
    assert!(ptr_0 < ptr_1);
}
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=093d1794ee25e921ba3dd819ec7a151b

File: compiler/stable_mir/src/ty.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (celinval) [reply compiler/stable_mir/src/ty.rs:290]: Is there any other function that you recommend removing?

[the remark this answers is not in the document]
Comment (compiler-errors) [reply compiler/stable_mir/src/ty.rs:290]: No, just this one

File: compiler/stable_mir/src/mir/body.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (celinval) [reply compiler/stable_mir/src/mir/body.rs:283]: and offset. :smile:

[the remark this answers is not in the document]
Comment (compiler-errors) [reply compiler/stable_mir/src/mir/body.rs:283]: Still could be done then?

**Comment by celinval** [comment]
@compiler-errors let me know if there's anything else you would like me to change.

**Review by compiler-errors — APPROVED** [review]
r=me

**Comment by workingjubilee** [comment]
> I don't think we lower SIMD intrinsics to MIR binops like we do for primitives, I can see it's not happening in `compiler/rustc_mir_transform/src/lower_intrinsics.rs`.

Correct, SIMD operations have their own path because they tend to be expressed as "don't think too hard about what's happening here in the MIR, this is just going to be lowered by the CG backend".

**Pull request merged**